version-ranges/src/lib.rs

```diff
@@ -1,122 +1,111 @@
-//!  2. Simplifying adjacent bounds of discrete sets cannot be detected and automated in the generic intersection code.
-//!
-//! An example for each can be given when `T` is `u32`.
-//! First, we can have both segments `S1 = (Unbounded, Included(42u32))` and `S2 = (Included(0), Included(42u32))`
-//! that represent the same segment but are structurally different.
-//! Thus, a derived equality check would answer `false` to `S1 == S2` while it's true.
-//!
-//! Second both segments `S1 = (Included(1), Included(5))` and `S2 = (Included(1), Included(3)) + (Included(4), Included(5))` are equal.
-//! But without asking the user to provide a `bump` function for discrete sets,
-//! the algorithm is not able to tell that the space between the right `Included(3)` bound and the left `Included(4)` bound is empty.
-//! Thus the algorithm is not able to reduce S2 to its canonical S1 form while computing sets operations like intersections in the generic code.
-//!
-//! This is likely to lead to user facing theoretically correct but practically nonsensical ranges,
-//! like (Unbounded, Excluded(0)) or (Excluded(6), Excluded(7)).
-//! In general nonsensical inputs often lead to hard to track bugs.
-//! But as far as we can tell this should work in practice.
-//! So for now this crate only provides an implementation for continuous ranges.
-//! With the v0.3 api the user could choose to bring back the discrete implementation from v0.2, as documented in the guide.
-//! If doing so regularly fixes bugs seen by users, we will bring it back into the core library.
-//! If we do not see practical bugs, or we get a formal proof that the code cannot lead to error states, then we may remove this warning.
+//! * `serde`: serialization and deserialization for the version range, given that the version type
+//!   also supports it.
+//! * `proptest`: Exports are proptest strategy for [`Ranges<u32>`].
 
 use std::borrow::Borrow;
 use std::cmp::Ordering;
 use std::fmt::{Debug, Display, Formatter};
 use std::ops::Bound::{self, Excluded, Included, Unbounded};
 use std::ops::RangeBounds;
 
-use crate::internal::SmallVec;
-use crate::VersionSet;
+#[cfg(any(feature = "proptest", test))]
+use proptest::prelude::*;
+use smallvec::{smallvec, SmallVec};
 
-/// A Range represents multiple intervals of a continuous range of monotone increasing
-/// values.
+/// Ranges represents multiple intervals of a continuous range of monotone increasing values.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize))]
 #[cfg_attr(feature = "serde", serde(transparent))]
-pub struct Range<V> {
-    segments: SmallVec<Interval<V>>,
+pub struct Ranges<V> {
+    /// Profiling in <https://github.com/pubgrub-rs/pubgrub/pull/262#discussion_r1804276278> showed
+    /// that a single stack entry is the most efficient. This is most likely due to `Interval<V>`
+    /// being large.
+    segments: SmallVec<[Interval<V>; 1]>,
```

Comment:
If this `1` ends up in practice to depend on the size of `V`, then we could make this a const generic on `Ranges`. Definitely premature optimization at this point.

Reply:
smallvec 2 uses const generics :eyes: